### DISCUSSION TITLE
Troubleshooting help

### DISCUSSION THREAD

**Discussion by jacobtsmith** [iplasvegas/uspto#21] [General]
Let's use this tab to discuss if you're cool with that. It will also allow me to have all info in one place for learning purposes.

**Comment by KevinJayData** [comment]
gotcha, this line 

`WebDriverWait(browser, 20).until(EC.visibility_of_element_located((By.XPATH, "//div[@class='expand_all expanded']"))).click()`

opens all of the tabs in this [link](https://tsdr.uspto.gov/#caseNumber=85931937&caseSearchType=US_APPLICATION&caseType=DEFAULT&searchType=statusSearch) so that the class names are searchable.  Otherwise selenium cannot find things under the tabs, i.e. `Standard Character Claim` underneath `Mark Information`.  Without running the above line you'll get a timeout error as selenium cuts itself off before looking for something forever.

**Reply by jacobtsmith** [reply]
I see. I added to a local branch and did notice that it opened all dropdowns. Still having this error though:

DevTools listening on ws://127.0.0.1:54522/devtools/browser/eb05e449-242f-44a5-812b-a122698d27b0
[2772:28276:0120/144837.709:ERROR:chrome_browser_main_extra_parts_metrics.cc(227)] START: ReportBluetoothAvailability(). If you don't see the END: message, this is crbug.com/1216328.
[2772:28276:0120/144837.710:ERROR:chrome_browser_main_extra_parts_metrics.cc(230)] END: ReportBluetoothAvailability()
[2772:28276:0120/144837.710:ERROR:chrome_browser_main_extra_parts_metrics.cc(235)] START: GetDefaultBrowser(). If you don't see the END: message, this is crbug.com/1216328.
[2772:3392:0120/144837.710:ERROR:device_event_log_impl.cc(214)] [14:48:37.711] USB: usb_service_win.cc:392 Could not read device interface GUIDs: The system cannot find the file specified. (0x2)
[2772:3392:0120/144837.712:ERROR:device_event_log_impl.cc(214)] [14:48:37.712] USB: usb_device_handle_win.cc:1050 Failed to read descriptor from node connection: A device attached to the system is not functioning. (0x1F)
[2772:28276:0120/144837.721:ERROR:chrome_browser_main_extra_parts_metrics.cc(239)] END: GetDefaultBrowser()
Traceback (most recent call last):
  File "C:\Users\jaket\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 187, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "C:\Users\jaket\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 110, in _get_module_details
    __import__(pkg_name)
  File "C:\clones\uspto\app.py", line 29, in <module>
    status = WebDriverWait(browser, 50).until(EC.visibility_of_element_located((By.XPATH, "//div[@class='key' and text()='Status']//following-sibling::div[1]"))).get_attribute("innerHTML")
  File "C:\clones\uspto\venv\lib\site-packages\selenium\webdriver\support\wait.py", line 89, in until
    raise TimeoutException(message, screen, stacktrace)
selenium.common.exceptions.TimeoutException: Message:
Stacktrace:
Backtrace:
        Ordinal0 [0x00D3FDC3+2555331]
        Ordinal0 [0x00CD77F1+2127857]
        Ordinal0 [0x00BD2E08+1060360]
        Ordinal0 [0x00BFE49E+1238174]
        Ordinal0 [0x00BFE69B+1238683]
        Ordinal0 [0x00C29252+1413714]
        Ordinal0 [0x00C17B54+1342292]
        Ordinal0 [0x00C275FA+1406458]
        Ordinal0 [0x00C17976+1341814]
        Ordinal0 [0x00BF36B6+1193654]
        Ordinal0 [0x00BF4546+1197382]
        GetHandleVerifier [0x00ED9622+1619522]
        GetHandleVerifier [0x00F8882C+2336844]
        GetHandleVerifier [0x00DD23E1+541697]
        GetHandleVerifier [0x00DD1443+537699]
        Ordinal0 [0x00CDD18E+2150798]
        Ordinal0 [0x00CE1518+2168088]
        Ordinal0 [0x00CE1660+2168416]
        Ordinal0 [0x00CEB330+2208560]
        BaseThreadInitThunk [0x76586739+25]
        RtlGetFullPathName_UEx [0x76FB8AFF+1215]
        RtlGetFullPathName_UEx [0x76FB8ACD+1165]

**Reply by KevinJayData** [reply]
Its because selenium can't find the `Status` key, this also occurs with `Earliest date §8 can be filed:` and I cannot find that one anywhere.  We'll need to hop on a call to find where those are at.

**Comment by jacobtsmith** [comment]
I also had a question about extracting an image since many marks also include an image mark.

**Reply by KevinJayData** [reply]
We cannot export the actual image into CSV but if you have access to excel then we can use the xlsxwriter package and skip csv in favor of excel.  https://stackoverflow.com/questions/41286889/how-to-insert-a-picture-into-excel-at-a-specified-cell-position-with-python-ana

- CSV = plain text only and we'd have to store the image as black and white pixel data since BW is two dimensional as opposed to color photos in RGB which is three dimensional (red, green, blue)
- excel = should allow for more complicated inputs.

**Reply by jacobtsmith** [reply]
That's what I was afraid of with CSV. It complicates things down the road with placing everything into a calendar (calendar takes in CSV's), but for the current client he really was strict on having images shown when there is an image mark.

**Comment by jacobtsmith** [comment]
Hey @KevinJayData 
I think I figured out whats happening. Not every serial will have every item we are searching for. Is there a way to create some sort of if statement that returns the value if it's found, but otherwise returns a null?

For example the original trademark we had been testing with isn't registered yet. Therefore when I tried to pull maintenance dates (I think like line 35) we crashed. Also, there may be slight variations between the naming. A section 8 and 9 declaration will take the same spot as a section 8 declaration, but be named accordingly.

**Reply by KevinJayData** [reply]
yeah python try/except clauses can do that where if an exception is raised then we can put a None/null value in its place. 

example looks like 
```
try:
    [selenium statement]
except Error as e:
    [save null as a value]
```

**Reply by jacobtsmith** [reply]
So, for example line 23 would it be 
`registration_date = try: [WebDriverWait(browser, 50).until(EC.visibility_of_element_located((By.XPATH, "//div[@class='key' and text()='Registration Date:']//following-sibling::div[1]"))).get_attribute("innerHTML")]
except Error as e:
[save null as a value]`

or am I off base here?

**Reply by jacobtsmith** [reply]
That looks sloppy I should have previewed before.
